### PR TITLE
Document what pyplot expects from a backend.

### LINKED ISSUE #12324: DOC: Write a unified backend doc
removed wx mocking in #12322 to unbreak the doc builc.  A todo would be to get this back in the docs in a more rational manner.

@dstansby 
> Is it possible to make wx.GetAp() return False in the mock object instead? It would be nice not to lose the docs...

@anntzer 
> We already got rid of qt and gtk docs for the same reason.  I'd rather write a proper unified interactive backends doc at some point...

_Originally posted by @anntzer in https://github.com/matplotlib/matplotlib/pull/12322#issuecomment-425471865_

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

(split out of #23101 to make that easier to go in first; goes on top of it.)

Closes #12324.

Now ready to go.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: doc/users/explain/writing_a_backend.rst [hunk 1]
```diff
@@ -0,0 +1,79 @@
+=================
+Writing a backend
+=================
+
+This page assumes general understanding of the information in the
+:doc:`/users/explain/backends` page, and is instead intended as reference for
+third-party backend implementers.  It also only deals with the interaction
+between backends and `.pyplot`, not with the rendering side, which is described
+in `.backend_template`.
```
Comment (timhoffm) [inline doc/users/explain/writing_a_backend.rst:9]: This split is somewhat inconvenient. Ok for a start, but in the long term, "writing a backend" should be summarized in one place. You could call this page "Interaction between backends and pyplot" and thus shift its focus.

File: doc/users/explain/writing_a_backend.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+Fundamentally, a backend module needs to provide information to `.pyplot`, so
+that
+
+1. `.pyplot.figure()` can create a new `.Figure` instance and associate it with
+   an instance of a backend-provided canvas class, itself hosted in an instance
+   of a backend-provided manager class.
+2. `.pyplot.show()` can show all figures and start the GUI event loop (if any).
+
+To do so, the backend module must define a class accessible as
+``backend_module.FigureCanvas``.  In the canvas-based API, this is the only
```
Comment (timhoffm) [inline doc/users/explain/writing_a_backend.rst:26]: Maybe add:
```suggestion
``backend_module.FigureCanvas``, that derives from `.FigureCanvasBase`.
In the canvas-based API, this is the only
```

File: doc/users/explain/writing_a_backend.rst [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+strict requirement for backend modules.  (The function-based API additionally
+requires many module-level functions to be defined.)
```
Comment (timhoffm) [inline doc/users/explain/writing_a_backend.rst:28]: No parentheses needed here
```suggestion
strict requirement for backend modules.  The function-based API additionally
requires many module-level functions to be defined.
```

File: doc/users/explain/writing_a_backend.rst [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+1. `.pyplot.figure()` calls
+   ``figure = Figure(); FigureCanvas.new_manager(figure, num)``
+   (``new_manager`` is a classmethod) to instantiate a canvas and a manager and
+   set up the ``figure.canvas`` and ``figure.canvas.manager`` attributes.
+   Figure unpickling uses the same approach, but replaces the newly
+   instantiated ``Figure()`` by the unpickled figure.
+
+   Interactive backends should customize the effect of ``new_manager`` by
+   setting a ``manager_class`` attribute on their ``FigureCanvas``, and
```
Comment (timhoffm) [inline doc/users/explain/writing_a_backend.rst:41]: setting it to what? The whole manager topic is missing.

File: doc/users/explain/writing_a_backend.rst [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+To do so, the backend module must define a class accessible as
+``backend_module.FigureCanvas``.  In the canvas-based API, this is the only
+strict requirement for backend modules.  (The function-based API additionally
+requires many module-level functions to be defined.)
+
+Canvas-based API (Matplotlib >= 3.6)
+------------------------------------
+
+1. `.pyplot.figure()` calls
```
Comment (timhoffm) [inline doc/users/explain/writing_a_backend.rst:33]:
```suggestion
1. **Creation of a figure**: `.pyplot.figure()` calls
```

File: doc/users/explain/writing_a_backend.rst [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+   ``FigureManager.create_with_canvas`` classmethod.  (Non-interactive backends
+   can normally use a trivial ``FigureManagerBase`` and can therefore skip this
+   step.)
+
+   After a new figure is registered with `.pyplot` (either via
+   `.pyplot.figure()` or via unpickling), if in interactive mode, `.pyplot`
+   will call its canvas' ``draw_idle()`` method, which can be overridden as
+   desired.
+
+2. `.pyplot.show()` calls ``FigureCanvas.manager_class.pyplot_show()`` (a
```
Comment (timhoffm) [inline doc/users/explain/writing_a_backend.rst:53]:
```suggestion
2. **Showing of a figure**: `.pyplot.show()` calls ``FigureCanvas.manager_class.pyplot_show()`` (a
```

File: doc/users/explain/writing_a_backend.rst [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+   By default, ``pyplot_show()`` checks whether there are any ``managers``
+   registered with `.pyplot` (exiting early if not), calls ``manager.show()``
+   on all such managers, and then, if called with ``block=True`` (or with
+   the default ``block=None`` and out of IPython's pylab mode and not in
+   interactive mode), calls ``FigureCanvas.manager_class.start_main_loop()``
+   (a classmethod) to start the main event loop.  Interactive backends should
+   therefore override the ``FigureCanvas.manager_class.start_main_loop``
+   classmethod accordingly (or alternatively, they may also directly override
+   ``FigureCanvas.manager_class.pyplot_show`` directly).
+
```
Comment (timhoffm) [inline doc/users/explain/writing_a_backend.rst:65]: I find this section a bit hard to follow as a recipe because it mixes explanations and things you have to do. But likely you need that interplay to understand what is going on.

Maybe additionally provide a tabular summary of the customizations / entry points.

**Review by timhoffm — COMMENTED** [review]
General note: You use parentheses for additional comments a lot. I suggest you try and reduce their use.

**Comment by anntzer** [comment]
Thanks for the review.  I've addressed all points except the last one, which I'm not really sure of how to handle and would suggest deferring.

**Review by timhoffm — APPROVED** [review]
